air/src/execution_step/instructions/ap_map.rs

```diff
@@ -0,0 +1,104 @@
+use crate::execution_step::execution_context::stream_map_key::StreamMapKey;
+use crate::execution_step::execution_context::StreamMapValueDescriptor;
+use crate::execution_step::instructions::ap::apply_to_arguments::apply_to_arg;
+use crate::execution_step::resolver::Resolvable;
+use crate::execution_step::ValueAggregate;
+use crate::log_instruction;
+use crate::trace_to_exec_err;
+use crate::CatchableError;
+use crate::ExecutionError;
+
+use air_interpreter_data::GenerationIdx;
+use air_parser::ast::ApMap;
+use air_parser::ast::ApMapKey;
+use air_parser::ast::Number;
+use air_parser::ast::StreamMap;
+use air_trace_handler::merger::MergerApResult;
+
+impl<'i> super::ExecutableInstruction<'i> for ApMap<'i> {
+    #[tracing::instrument(level = "debug", skip(exec_ctx, trace_ctx))]
+    fn execute(&self, exec_ctx: &mut ExecutionCtx<'i>, trace_ctx: &mut TraceHandler) -> ExecutionResult<()> {
+        log_instruction!(call, exec_ctx, trace_ctx);
+        // this applying should be at the very beginning of this function,
+        // because it's necessary to check argument lambda, for more details see
+        // https://github.com/fluencelabs/aquavm/issues/216
+        let result = apply_to_arg(&self.value, exec_ctx, trace_ctx, true)?;
+
+        let merger_ap_result = to_merger_ap_map_result(&self, trace_ctx)?;
+        let generation = populate_context(&self.key, &self.map, &merger_ap_result, result, exec_ctx)?;
+        maybe_update_trace(generation, trace_ctx);
+
+        Ok(())
+    }
+}
+
+fn to_merger_ap_map_result(instr: &impl ToString, trace_ctx: &mut TraceHandler) -> ExecutionResult<MergerApResult> {
+    let merger_ap_result = trace_to_exec_err!(trace_ctx.meet_ap_start(), instr)?;
+    Ok(merger_ap_result)
+}
+
+fn populate_context<'ctx>(
+    key: &ApMapKey<'ctx>,
+    ap_map_result: &StreamMap<'ctx>,
+    merger_ap_result: &MergerApResult,
+    result: ValueAggregate,
+    exec_ctx: &mut ExecutionCtx<'ctx>,
+) -> ExecutionResult<GenerationIdx> {
+    let value_descriptor = generate_map_value_descriptor(result, ap_map_result, merger_ap_result);
+    match key {
+        ApMapKey::Literal(s) => exec_ctx
+            .stream_maps
+            .add_stream_map_value(&s.to_owned().into(), value_descriptor),
+        ApMapKey::Number(n) => match n {
+            Number::Int(int) => exec_ctx
+                .stream_maps
+                .add_stream_map_value(&int.to_owned().into(), value_descriptor),
+            Number::Float(_) => Err(CatchableError::StreamMapError(FloatMapKeyIsUnsupported {
+                variable_name: String::from(ap_map_result.name),
+            })
+            .into()),
+        },
+        ApMapKey::Scalar(s) => resolve_add(s, exec_ctx, value_descriptor, ap_map_result.name),
+        ApMapKey::ScalarWithLambda(s) => resolve_add(s, exec_ctx, value_descriptor, ap_map_result.name),
+        ApMapKey::CanonStreamWithLambda(c) => resolve_add(c, exec_ctx, value_descriptor, ap_map_result.name),
+    }
+}
+
+fn resolve_add(
```

Comment:
what does `add` here mean?

Reply:
Oh, it's for addition, it's better to do it more granularly and not in one function.